### PR TITLE
feat: CHK-2625 add warmup

### PR DESCRIPTION
**Pull request by a-mannia**
#### List of Changes
Speed up the launch of pods into production by adding warmup.

#### Motivation and Context
The goal of this PR is to speedup the launch of pods

#### How Has This Been Tested?
Unit test

#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

Depends on pagopa/pagopa-ecommerce-local#163

### CODE REVIEW

File: src/main/java/it/pagopa/ecommerce/payment/methods/controller/PaymentMethodsController.java [hunk 1]
```diff
@@ -289,4 +292,16 @@ private Mono<String> getAuthenticationToken(ServerWebExchange exchange) {
                         .map(header -> header.substring("Bearer ".length()))
         );
     }
+
+    @Warmup
```
Comment (pietro-tota) [inline src/main/java/it/pagopa/ecommerce/payment/methods/controller/PaymentMethodsController.java:296]: What about adding also a method to perform warmup for a calculate fees endpoint too? 
This should be a safe operation to be performed as warmup requests.

cc @infantesimone

Comment (a-mannia) [reply src/main/java/it/pagopa/ecommerce/payment/methods/controller/PaymentMethodsController.java:297]: done

**Comment by pietro-tota** [comment]
please resolve conflicts :D

File: src/main/java/it/pagopa/ecommerce/payment/methods/controller/PaymentMethodsController.java [hunk 2]
```diff
@@ -289,4 +296,53 @@ private Mono<String> getAuthenticationToken(ServerWebExchange exchange) {
...
+                .get()
+                .uri("http://localhost:8080/payment-methods")
+                .header(X_CLIENT_ID, PaymentMethodRequestDto.ClientIdEnum.CHECKOUT.toString())
+                .retrieve()
+                .bodyToMono(PaymentMethodsResponseDto.class)
+                .block(Duration.ofSeconds(30));
+    }
+
+    @Warmup
+    public void calculateFeesWarmupMethod() {
```
Comment (pietro-tota) [inline src/main/java/it/pagopa/ecommerce/payment/methods/controller/PaymentMethodsController.java:313]: could you please add this warmup method also for v2 controller too?

Comment (a-mannia) [reply src/main/java/it/pagopa/ecommerce/payment/methods/controller/PaymentMethodsController.java]: Done.

File: src/main/java/it/pagopa/ecommerce/payment/methods/controller/v2/PaymentMethodsController.java [hunk 3]
```diff
@@ -74,4 +83,27 @@ public ResponseEntity<ProblemJsonDto> errorHandler(RuntimeException exception) {
...
 
+    @Warmup
+    public void calculateFeesWarmupMethod() {
+        CalculateFeeRequestDto request = new CalculateFeeRequestDto()
+                .bin("BIN_TEST")
+                .touchpoint("CHECKOUT")
+                .addIdPspListItem("string")
+                .idPspList(new ArrayList<>(List.of("first", "second")))
+                .isAllCCP(false)
+                .paymentNotices(Collections.emptyList());
```
Comment (pietro-tota) [inline src/main/java/it/pagopa/ecommerce/payment/methods/controller/v2/PaymentMethodsController.java:94]:
```suggestion
          CalculateFeeRequestDto request = new CalculateFeeRequestDto()
                .touchpoint("CHECKOUT")
                .isAllCCP(false)
                .addPaymentNoticesItem(
                        new PaymentNoticeDto()
                                .paymentAmount(100L)
                                .primaryCreditorInstitution("77777777777")
                                .addTransferListItem(
                                        new TransferListItemDto()
                                                .digitalStamp(false)
                                                .creditorInstitution("77777777777")
                                )

                );
```
The input request is invalid (warmup call lead to a 400 BAD request.
Change it as follow to produce a valid request

File: src/main/java/it/pagopa/ecommerce/payment/methods/controller/v1/PaymentMethodsController.java [hunk 4]
```diff
@@ -289,4 +296,53 @@ private Mono<String> getAuthenticationToken(ServerWebExchange exchange) {
...
+    }
+
+    @Warmup
+    public void createSessionWarmupMethod() {
+        WebClient
+                .create()
+                .post()
+                .uri(
+                        "http://localhost:8080/payment-methods/{id}/sessions",
+                        UUID.randomUUID().toString()
```
Comment (pietro-tota) [inline src/main/java/it/pagopa/ecommerce/payment/methods/controller/v1/PaymentMethodsController.java:341]: please add a new env variable from which read the payment method id to be used during this warmup api call in order to make warmup request to be processed successfully.
Payment method ids to be used are (based on the env):

- DEV -> e7058cac-5e1a-4002-8994-5bab31e9f385
- UAT -> 378d0b4f-8b69-46b0-8215-07785fe1aad4
- PROD -> 83b5e0c7-2d35-44f4-9e48-a168b84433a1

File: src/main/java/it/pagopa/ecommerce/payment/methods/controller/v2/PaymentMethodsController.java [hunk 5]
```diff
@@ -74,4 +83,27 @@ public ResponseEntity<ProblemJsonDto> errorHandler(RuntimeException exception) {
...
+        WebClient
+                .create()
+                .post()
+                .uri(
+                        "http://localhost:8080/v2/payment-methods/{id}/fees",
+                        UUID.randomUUID().toString()
```
Comment (pietro-tota) [inline src/main/java/it/pagopa/ecommerce/payment/methods/controller/v2/PaymentMethodsController.java:100]: same as above

File: src/main/java/it/pagopa/ecommerce/payment/methods/controller/v1/PaymentMethodsController.java (hunk 4 above)
Comment (a-mannia) [reply src/main/java/it/pagopa/ecommerce/payment/methods/controller/v1/PaymentMethodsController.java]: Thanks for the ids. Done.

File: src/main/java/it/pagopa/ecommerce/payment/methods/controller/v2/PaymentMethodsController.java (hunk 5 above)
Comment (a-mannia) [reply src/main/java/it/pagopa/ecommerce/payment/methods/controller/v2/PaymentMethodsController.java]: Done.

File: .devops/code-review-pipelines.yml [hunk 6]
```diff
@@ -89,7 +89,7 @@ stages:
               echo "##[debug] Change placeholder for ecommerce-payment-method-service set branch: $(BRANCH_NAME)"
               sed -i "s/^ECOMMERCE_PAYMENT_METHODS_COMMIT_SHA=.*$/ECOMMERCE_PAYMENT_METHODS_COMMIT_SHA=$(BRANCH_NAME)/" .env
             displayName: 'Change payment-method branch name ecommerce-local .env'
-          - script: docker compose up pagopa-ecommerce-payment-methods-service pagopa-gec-mock pagopa-npg-mock mongo redis -d
+          - script: docker compose up pagopa-ecommerce-payment-methods-service pagopa-gec-mock pagopa-npg-mock mongo redis
```
Comment (pietro-tota) [inline .devops/code-review-pipelines.yml:92]:
```suggestion
          - script: docker compose up pagopa-ecommerce-payment-methods-service pagopa-gec-mock pagopa-npg-mock mongo redis -d
```
restore this command

File: .env.example [hunk 7]
```diff
@@ -15,23 +15,26 @@ REDIS_HOST=localhost
...
 OTEL_RESOURCE_ATTRIBUTES="service.name=pagopa-ecommerce-payment-methods-service,deployment.environment=dev"
 OTEL_EXPORTER_OTLP_ENDPOINT="http://otel-collector.elastic-system.svc:4317"
 OTEL_LOGS_EXPORTER="none"
 OTEL_TRACES_SAMPLER="always_on"
 OTEL_EXPORTER_OTLP_HEADERS="Authorization=Bearer aaa"
 NPG_NOTIFICATION_JWT_SECRET=ODMzNUZBNTZENDg3NTYyREUyNDhGNDdCRUZDNzI3NDMzMzQwNTFEREZGQ0MyQzA5Mjc1RjY2NTQ1NDk5MDMxNzU5NDc0NUVFMTdDMDhGNzk4Q0Q3RENFMEJBODE1NURDREExNEY2Mzk4QzFEMTU0NTExNjUyMEExMzMwMTdDMDk
-NPG_NOTIFICATION_JWT_VALIDITY_TIME=900
\ No newline at end of file
+NPG_NOTIFICATION_JWT_VALIDITY_TIME=900
+WARMUP_PAYMENT_METHOD_ID="e7058cac-5e1a-4002-8994-5bab31e9f385"
```
Comment (pietro-tota) [inline .env.example:40]: please open a PR against ecommerce-local repo updating [payment methods service env variables](https://github.com/pagopa/pagopa-ecommerce-local/blob/main/pagopa-ecommerce-payment-methods-service/payment-methods-service.env)

Comment (giovanniberti) [reply .env.example:40]: Also add the variable to `README.md` if you haven't already :)

Comment (a-mannia) [reply .env.example:40]: Thank you guys. Done.

File: .devops/code-review-pipelines.yml (hunk 6 above)
Comment (a-mannia) [reply .devops/code-review-pipelines.yml]: Done.

**Review by pietro-tota — APPROVED** [review]

**Pull request merged**